gpii/node_modules/lifecycleManager/src/UserLogonRequest.js

```diff
@@ -0,0 +1,228 @@
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/GPII/universal/blob/master/LICENSE.txt
+ */
+
+"use strict";
+
+var fluid = fluid || require("infusion"),
+    gpii = fluid.registerNamespace("gpii");
+
+/** One dynamic component of this type is constructed for each user logon request on the LifecycleManager.
+ */
+fluid.defaults("gpii.lifecycleManager.userLogonRequest", {
+    gradeNames: ["fluid.component"],
+    gpiiKey: null,          // A GPII key, must be supplied by integrators
+    requestPromise: null,   // A promise, must be supplied by integrators
+    events: {
+        onSuccess: null,
+        onError: null
+    },
+    invokers: {
+        handleRequest: {
+            funcName: "gpii.lifecycleManager.userLogonRequest.handleRequest",
+            args: ["{that}", "{lifecycleManager}", "{arguments}.0"]
+                                                   // logonState
+        }
+    },
+    listeners: {
+        "onSuccess.handleSuccessRequest": {
+            listener: "gpii.lifecycleManager.userLogonRequest.handleSuccessRequest",
+            args: ["{that}", "{arguments}.0"]
+        },
+        "onError.handleFailedRequest": {
+            listener: "gpii.lifecycleManager.userLogonRequest.handleFailedRequest",
+            args: ["{that}", "{arguments}.0"],
+            priority: "after:handleErrors"
+        }
+    }
+});
+
+gpii.lifecycleManager.userLogonRequest.handleRequest = function (that, lifecycleManager, logonState) {
+    var gpiiKey = that.options.gpiiKey;
+    var activeGpiiKeys = lifecycleManager.getActiveSessionGpiiKeys(); // find currently logged in user:
+
+    // keep track of the last non-noUser key for /proximityTriggered endpoint to verify its debounce rule
+    // when the the current active key is "noUser".
+    if (gpiiKey !== "noUser") {
+        gpii.lifecycleManager.userLogonHandling.updateLastNonNoUserLogon(lifecycleManager, gpiiKey, logonState);
+    }
+
+    // handle reset requests
+    if (gpiiKey === "reset") {
+        that.handleResetGpiiKey();
+    }
+
+    // handle keyin requests
+    if (logonState === "login") {
+        // if another user, excluding "noUser", is already logged in
+        if (activeGpiiKeys.length === 1 && activeGpiiKeys[0] !== "noUser") {
+            that.errorResponse("Got log in request from user " + gpiiKey + ", but the user " +
+                activeGpiiKeys[0] + " is already logged in. So ignoring login request.", 409);
+        } else {
+            that.loginUser(gpiiKey);
+        }
+    }
+
+    // handle keyout requests
+    if (logonState === "logout") {
+        if (activeGpiiKeys[0] !== gpiiKey) {
+            that.errorResponse("Got logout request from user " + gpiiKey + ", but the user " +
+                activeGpiiKeys[0] + " is logged in. So ignoring the request.", 409);
+        } else {
+            that.logoutUser(gpiiKey);
+        }
+    }
+
+    // handle getGpiiKey requests
+    if (logonState === "getGpiiKey") {
```

Comment:
I don't think the extra complexity of this queue "action" type is needed - we always know synchronously who is logged on. This can just become a simple invoker on the component. Note that withSession is i) always synchronous, ii) will be eliminated in the next refactoring once we remove the possibility for multiple active sessions.